src/app/blog/blob-store-design-challenges/page.mdx

```diff
@@ -0,0 +1,215 @@
+import { BlogPostLayout } from '@/components/BlogPostLayout'
+import {ThemeImage} from '@/components/ThemeImage'
+
+export const post = {
+  draft: false,
+  author: 'Rüdiger Klaehn',
+  date: '2025-02-24',
+  title: 'Blob store design challenges',
+  description:
+    "It's just hashed data, what's so complex about a blob store anyway?",
+}
+
+export const metadata = {
+  title: post.title,
+  description: post.description,
+}
+
+export default (props) => <BlogPostLayout article={post} {...props} />
+
+We've been hard at work with one of our protocols that you can pick off the shelf with iroh: [iroh-blobs]
+You might've used iroh-blobs before if you tried out [`sendme`](https://www.iroh.computer/sendme), which is just a small CLI wrapper over iroh with iroh-blobs.
+
+Put briefly, iroh-blobs helps transfer arbitrary data ("blobs") from one device to another via iroh in a way that allows these transfers to be interrupted, resumed and verified by content-addressing blobs using the BLAKE3 hashing function.
+
+Iroh-blobs also allows you to exchange the backing storage it uses via *blob store* abstraction, so the actual hashes and blobs might be stored as files on disk, stored fully in-memory or even fully remote by utilizing some cloud storage.
+
+Now, BLAKE3 hashed data and BLAKE3 verified streaming/bao are fully deterministic.
+
+The iroh blobs protocol is implemented in this spirit. Even for complex requests like requests that involve multiple ranges or entire hash sequences, for every request, there is exactly one sequence of bytes that is the correct answer. And the requester will notice if *data* is incorrect after at most 16 KiB of data.
+
+So why is designing a blob store for BLAKE3 hashed data so complex? What are the challenges?
+
+## Terminology
+
+The job of a blob store is to store two pieces of data per BLAKE3 `hash`, the actual data itself and an `outboard` containing the flattened hash tree that connects each `chunk` of data to the root hash. Data and outboard are kept separate so that the data can be used as-is.
+
+While the [bao] crate stores tree hashes all the way down to &le; 1024 byte BLAKE3 `chunks`, we store the outboard only down to 16 KiB `chunk groups` to reduce the outboard size. 
+
+## In memory store
+
+In the simplest case, you store data in memory. If you want to serve complete blobs, you just need a map from `hash` to `data` and `outboard` stored in a cheaply cloneable container like [bytes::Bytes]. Even if you want to also handle incomplete blobs, you use a map from `hash` and `chunk number` to (&le; 1024 byte) data block, and that's it. Such a store is relatively easy to implement using a crate like [bao-tree] and a simple [BTreeMap] wrapped in a [Mutex].
```

Comment:
```suggestion
In the simplest case, you store data in memory. If you want to serve complete blobs, you just need a map from `hash` to `data` and `outboard` stored in a cheaply cloneable container like [`bytes::Bytes`]. Even if you want to also handle incomplete blobs, you use a map from `hash` and `chunk number` to (&le; 1024 byte) data block, and that's it. Such a store is relatively easy to implement using a crate like [bao-tree] and a simple [`BTreeMap`] wrapped in a [`Mutex`].
```

Reply:
Yeah, I don't like it. For me it is putting styling information into markdown, and I think the cost benefit ratio for blog posts where you should be able to write them quickly is not there. But... 🤷‍♂️ ...